---
title: "snRNAseq workflow: dimensionality reduction and clustering"
date: "`r format(Sys.time(), '%B %d, %Y')`"
geometry: margin=1in
fontfamily: mathpazo
fontsize: 11pt
params:
  cache_dir: false
output:
  html_document:
    keep_md: true
    code_folding: show
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    number_sections: true
    theme: lumen
---

```{r initialise, include = F}
# load packages
library(SCPA)
library(msigdbr)
library(magrittr)
library(dplyr)
library(SummarizedExperiment)
library(tidyverse)
library(monocle3)

# set ggplot presets
ggplot2::theme_set(ggplot2::theme_bw())
ggplot2::theme_update(legend.text = ggplot2::element_text(size = 7))
ditto_colours <- list(ggplot2::scale_fill_manual(values = dittoSeq::dittoColors(), na.value = "white"),
                      ggplot2::scale_colour_manual(values = dittoSeq::dittoColors(), na.value = "white"))
umap_void_theme <-
  ggplot2::theme(axis.text = ggplot2::element_blank(),
                 axis.title = ggplot2::element_blank(),
                 axis.ticks = ggplot2::element_blank(),
                 legend.position = "none",
                 plot.margin = ggplot2::unit(c(2,2,2,2), "pt"))

# chunk setup
knitr::opts_chunk$set(fig.align = "center",
                      dpi = 300,
                      cache.path = params$cache_dir)

# functions
fix_pathway_names <- function(pathway_names) {
  pathway_names %>% tolower() %>% gsub("\\_", " ", .)
}
plot_my_heatmap <- function(p_dat_in, col, stat = "qval") {
  p_dat_in %>%
    select(Pathway, name = matches(col), value = matches(stat)) %>%
    mutate(Pathway = fix_pathway_names(Pathway)) %>%
    pivot_wider() %>%
    column_to_rownames("Pathway") %>%
    as.matrix() %>%
    Heatmap(name = stat,
            border = T,
            show_row_dend = F,
            show_column_dend = T,
            row_names_gp = grid::gpar(fontsize = 8))
}
plot_my_enrichment <- function(p_dat_in, top_n = 5) {
  p_dat <-
    p_dat_in %>%
    mutate(Pathway = Pathway %>% fix_pathway_names(),
           color = case_when(FC > 5 & adjPval < 0.01 ~ '#6dbf88',
                             FC < 5 & FC > -5 & adjPval < 0.01 ~ '#84b0f0',
                             FC < -5 & adjPval < 0.01 ~ 'mediumseagreen',
                             FC < 5 & FC > -5 & adjPval > 0.01 ~ 'black'))
  ggplot(p_dat, aes(-FC, qval)) +
    geom_vline(xintercept = c(-5, 5), linetype = "dashed", col = 'black', lwd = 0.3) +
    geom_point(cex = 2.6, shape = 21, fill = p_dat$color, stroke = 0.3) +
    ggrepel::geom_text_repel(
      data = p_dat %>% filter(
        (rank(FC, ties.method = "min") <= 5 | rank(-qval, ties.method = "min") <= 5) &
        abs(FC) > 5 & adjPval < 0.01), 
      aes(label = Pathway),
                             size = 4, min.segment.length = 0, nudge_x = 10) +
    xlab("Enrichment") +
    ylab("Qval") +
    theme(panel.background = element_blank(),
          panel.border = element_rect(fill = NA),
          aspect.ratio = 1) +
    ggplot2::theme(text = ggplot2::element_text(size = 20))
}
find_markers <- function(seu, col, group.1, group.2) {
  markers <-
    seu %>% 
    Seurat::FindMarkers(
      ident.1 = colnames(seu[, seu@meta.data[, col] == group.1]),
      ident.2 = colnames(seu[, seu@meta.data[, col] == group.2]),
      test.use = "MAST"
    ) %>%
    tibble::as_tibble(rownames = "gene") 
  
  if (nrow(markers) > 0 & "p_val_adj" %in% colnames(markers)) {
    markers <-
      markers %>%
      dplyr::mutate(direction = ifelse(avg_log2FC > 0, group.1, group.2),
                    group = ifelse(avg_log2FC > 0, 1, 2)) %>%
      dplyr::group_by(direction) %>%
      dplyr::mutate(rank = rank(-abs(avg_log2FC)),
                    p_val_adj_no_0 = dplyr::case_when(p_val_adj == 0 ~ 2.225074e-308,
                                                     TRUE ~ p_val_adj))
  } else {
    message("No differentially expressed genes!")
    markers <- tibble::tibble()
  }
  markers
}
plot_markers <- function(markers, top_n = 10) {
  p_dat <-
    markers %>%
    dplyr::filter(rank <= top_n) %>%
    dplyr::group_by(direction) %>%
    dplyr::mutate(rank = rank(-abs(avg_log2FC))) %>%
    dplyr::filter(rank <= top_n)
  p_dat %>%
    ggplot(aes(x = -rank, y = avg_log2FC, 
               fill = direction, alpha = -log10(p_val_adj_no_0),
               label = gene)) +
    geom_bar(stat = "identity") +
    geom_text(hjust = ifelse(p_dat$group == 2, 1.1, -0.1),
              alpha = 1,
              size = 6) +
    scale_y_continuous(labels = abs, limits = max(abs(p_dat$avg_log2FC)) * c(-1,1) * 1.3) +
    scale_fill_manual(values=as.vector(c("#3179de", "#d23f67"))) +
    scale_alpha(range = c(0.5, 1)) +
    labs(x = "", y = "", fill="") +
    theme_minimal() +   
    coord_flip() +
    theme( 
      panel.grid.major = element_blank(), 
      panel.grid.minor = element_blank(),
      axis.text.x = element_text(size=20), 
      axis.text.y = element_blank(),
      strip.text.x=element_text(size=24),
      legend.position="bottom",
      legend.text=element_text(size=20)
    )
}
```

# Differential expression analysis

```{r load_seu}
# load seu 
seu <- readRDS("seu.rds")
Seurat::DefaultAssay(seu) <- "RNA"
```

```{r find_markers}
degs <- list()
degs <- xfun::cache_rds({

  if (all(c("Malignant_cells", "Epithelial_cells") %in% seu$label.infercnv)) {
    # find markers of malignant vs epithelial
    degs[["malignant_vs_epithelial"]] <-
       find_markers(seu, "label.infercnv", "Malignant_cells", "Epithelial_cells")
  }
  
  # if (all(c("pre", "post") %in% seu[, seu$label.infercnv == "Malignant_cells"]$timepoint)) {
  #   # find markers of malignant pre vs malignant post
  #   degs[["malignant_post_vs_malignant_pre"]] <-
  #      find_markers(seu[, seu$label.infercnv == "Malignant_cells"], "timepoint", "post", "pre")
  # }
  # 
  # if (all(c("lymph_node", "tumour") %in% seu[, seu$label.infercnv == "Malignant_cells"]$sample_site)) {
  #   # find markers of malignant lymph node vs malignant primary
  #   degs[["malignant_lymph_node_vs_malignant_tumour"]] <-
  #      find_markers(seu[, seu$label.infercnv == "Malignant_cells"], "sample_site", "lymph_node", "tumour")
  # }
  
  # return
  degs
}, file = "degs.rds")
```

```{r plot_degs, eval = F}
# plot
plot_markers(degs$malignant_vs_epithelial, 20)
# plot_markers(degs$malignant_post_vs_malignant_pre, 20)
# plot_markers(degs$post_vs_pre, 20)
```

# Pathway analysis

```{r pathway_analysis, eval = F}
# convert to cds
cds <- SeuratWrappers::as.cell_data_set(seu, assay = "RNA")
SummarizedExperiment::rowData(cds)$gene_short_name <- 
  row.names(SummarizedExperiment::rowData(cds))

# preprocess
cds <- preprocess_cds(cds, num_dim = 30)
cds <- reduce_dimension(cds)

# get hallmark pathways
pathways <- xfun::cache_rds({
  pathways <-
    dplyr::bind_rows(
      msigdbr(species = "Homo sapiens", category = "H"),
      msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:KEGG"),
      msigdbr(species = "Homo sapiens", category = "C2", subcategory = "CP:REACTOME")
    ) %>%
    format_pathways()
  pathways
}, file = "pathways.rds")

# compare pathways between tumour vs normal
scpa_out <- list()
scpa_out <- xfun::cache_rds({
  scpa_out[["malignant_vs_epithelial"]] <-
    compare_sce(cds,
                group1 = "label.infercnv",
                group1_population = c("Malignant_cells", "Epithelial_cells"),
                pathways = pathways)
  scpa_out
}, file = "scpa.rds")
```

```{r plot_pathways_dot, eval = F}
# plot
plot_my_enrichment(scpa_out$malignant_vs_epithelial)
# plot_my_enrichment(scpa_out$malignant_post_vs_malignant_pre)
# plot_my_enrichment(scpa_out$post_vs_pre)
```

```{r plot_pathways_heatmap, eval = F}
# plot
plot_heatmap(scpa_out$malignant_vs_epithelial %>% filter(adjPval < 0.01))
```
