---
title: "snRNAseq QC report"
date: "`r format(Sys.time(), '%B %d, %Y')`"
geometry: margin=1in
fontfamily: mathpazo
fontsize: 11pt
params:
  params_file: false
output:
  html_document:
    keep_md: true
    code_folding: show
    toc: true
    toc_float: true
    toc_collapsed: true
    toc_depth: 4
    number_sections: true
    theme: lumen
---

```{r initialise, include = F}
# magrittr, gridExtra, clustree, monocle3 must be loaded in the environment (due to package-specific bugs)
library(clustree)
library(gridExtra)
library(monocle3)
library(magrittr)

# load params file
# base_dir=ifelse(Sys.info()["nodename"]=="Alexs-MacBook-Air-2.local","/Volumes/TracerX/","/camp/project/tracerX/");setwd(paste0(base_dir,"working/VHL_GERMLINE/tidda/snRNAseq_workflow/"))
# cluster: # params <- list(params_file = "output/params.json")
# local:   # params <- list(params_file = "test/local_params.json")
curr_params <- 
  rjson::fromJSON(file = params$params_file) %>% 
  type.convert(as.is = T)

# set seed for repeatability
set.seed(42)
random_seed = 42

# set ggplot presets
ggplot2::theme_set(ggplot2::theme_bw())
ggplot2::theme_update(legend.text = ggplot2::element_text(size = 7))
ditto_colours <- list(ggplot2::scale_fill_manual(values = dittoSeq::dittoColors(), na.value = "white"),
                      ggplot2::scale_colour_manual(values = dittoSeq::dittoColors(), na.value = "white"))
umap_void_theme <-
  ggplot2::theme(axis.text = ggplot2::element_blank(),
                 axis.title = ggplot2::element_blank(),
                 axis.ticks = ggplot2::element_blank(),
                 legend.position = "none",
                 plot.margin = ggplot2::unit(c(2,2,2,2), "pt"))

# chunk setup
knitr::opts_chunk$set(fig.align = "center",
                      dpi = 300)

# set sample groupings to check at the clustering stage
groupings <-
  c("sample", "percent_mito", "percent_ribo", "percent_globin",
    "date_prep", "nih_pid", "rin", "lesion_type", "tumour_size", "fuhrman_grade") %>%
  # if genome is human, do cell cycle scoring (doesn't work with other genomes)
  { if (grepl("hg38", curr_params$genome)) c(., "Phase") else . } %>%
  # if checking for doublets, add to the groupings
  { if (curr_params$filter$doublets == T) c(., "doublet") else . } 
```

```{r define_functions, include = F, eval = T}
# flatten lists
flatten_list <- function (x, use.names = TRUE, classes = "ANY") 
{
    len <- sum(rapply(x, function(x) 1L, classes = classes))
    y <- vector("list", len)
    i <- 0L
    items <- rapply(x, function(x) {
        i <<- i + 1L
        y[[i]] <<- x
        TRUE
    }, classes = classes)
    if (use.names && !is.null(nm <- names(items))) 
        names(y) <- nm
    y
}

# calculate plot height for grids
get_fig_dims <- function(n_plots, n_cols = NULL, grid_width = 10, height_to_width_ratio = 1) {
  # get dims
  dims <- ggplot2::wrap_dims(n_plots, n_cols)
  # get scale
  scale <- grid_width / dims[1]
  # get height
  grid_height <- (dims[2] * scale) * height_to_width_ratio
  # return dims
  c(grid_width, grid_height)
}

# Get cluster centroids from a Seurat or cell_data_set object
get_centroids <- function(object, reduction, lvl) {

  if (class(object)[1] == "Seurat") {
    embeddings <- object@reductions[[reduction]]@cell.embeddings
    metadata <- object@meta.data
  } else if (class(object)[1] == "cell_data_set") {
    embeddings <- SingleCellExperiment::reducedDims(object)[[reduction]]
    metadata <- SummarizedExperiment::colData(object) %>% dplyr::as_tibble()
  }

  dplyr::tibble(
    x = embeddings[,1],
    y = embeddings[,2],
    metadata
  ) %>%
    dplyr::group_by(get(lvl)) %>%
    dplyr::summarise(x = median(x), y = median(y)) %>%
    dplyr::rename(!!lvl := `get(lvl)`)

}

#' Identify the genes most specifically expressed in groups of cells
#'
#' @param cds A cell_data_set object to calculate top markers for.
#' @param group_cells_by String indicating what to group cells by for
#'   comparison. Default is "cluster".
#' @param genes_to_test_per_group Numeric, how many genes of the top ranked
#'   specific genes by Jenson-Shannon to do the more expensive regression test
#'   on.
#' @param reduction_method String indicating the method used for dimensionality
#'   reduction. Currently only "UMAP" is supported.
#' @param marker_sig_test A flag indicating whether to assess the discriminative
#' power of each marker through logistic regression. Can be slow, consider
#' disabling to speed up top_markers().
#' @param reference_cells If provided, top_markers will perform the marker
#' significance test against a "reference set" of cells. Must be either a list
#' of cell ids from colnames(cds), or a positive integer. If the latter, top_markers()
#' will randomly select the specified number of reference cells. Accelerates
#' the marker significance test at some cost in sensitivity.
#' @param speedglm.maxiter Maximum number of iterations allowed for fitting GLM
#' models when testing markers for cell group.
#' @param cores Number of cores to use.
#' @param verbose Whether to print verbose progress output.
#'
#' @return a data.frame where the rows are genes and the columns are
#' * gene_id vector of gene names
#' * gene_short_name vector of gene short names
#' * cell_group character vector of the cell group to which the cell belongs
#' * marker_score numeric vector of marker scores as the fraction expressing scaled by the specificity. The value ranges from 0 to 1.
#' * mean_expression numeric vector of mean normalized expression of the gene in the cell group
#' * fraction_expressing numeric vector of fraction of cells expressing the gene within the cell group
#' * specificity numeric vector of a measure of how specific the gene's expression is to the cell group based on the Jensen-Shannon divergence. The value ranges from 0 to 1.
#' * pseudo_R2 numeric vector of pseudo R-squared values, a measure of how well the gene expression model fits the categorical data relative to the null model. The value ranges from 0 to 1.
#' * marker_test_p_value numeric vector of likelihood ratio p-values
#' * marker_test_q_value numeric vector of likelihood ratio q-values
#'
#' @examples
#'   \donttest{
#'     library(dplyr)
#'
#'     cell_metadata <- readRDS(system.file('extdata',
#'                                          'worm_embryo/worm_embryo_SummarizedExperiment::colData.rds',
#'                                          package='monocle3'))
#'     gene_metadata <- readRDS(system.file('extdata',
#'                                          'worm_embryo/worm_embryo_SummarizedExperiment::rowData.rds',
#'                                          package='monocle3'))
#'     expression_matrix <- readRDS(system.file('extdata',
#'                                              'worm_embryo/worm_embryo_expression_matrix.rds',
#'                                              package='monocle3'))
#'
#'     cds <- new_cell_data_set(expression_data=expression_matrix,
#'                              cell_metadata=cell_metadata,
#'                              gene_metadata=gene_metadata)
#'
#'     cds <- preprocess_cds(cds)
#'     cds <- reduce_dimension(cds)
#'     cds <- cluster_cells(cds)
#'     marker_test_res <- top_markers(cds, group_cells_by="partition", reference_cells=1000)
#'     top_specific_markers <- marker_test_res %>%
#'                              filter(fraction_expressing >= 0.10) %>%
#'                              group_by(cell_group) %>%
#'                              top_n(1, pseudo_R2)
#'     top_specific_marker_ids <- unique(top_specific_markers %>% pull(gene_id))
#'   }
#'
#' @importFrom dplyr n
#' @export
top_markers_develop <- function(cds,
                        group_cells_by="cluster",
                        genes_to_test_per_group=25,
                        reduction_method="UMAP",
                        marker_sig_test=TRUE,
                        reference_cells=NULL,
                        speedglm.maxiter=25,
                        cores=1,
                        verbose=FALSE) {
  # testing: # group_cells_by="partition";genes_to_test_per_group=25;reduction_method="UMAP";marker_sig_test=TRUE;reference_cells=NULL;speedglm.maxiter=25;cores=1;verbose=FALSE
  rowname <- cell_group <- marker_score <- cell_id <- mean_expression <- NULL # no visible binding
  fraction_expressing <- specificity <- pseudo_R2 <- NULL # no visible binding
  lrtest_p_value <- lrtest_q_value <- gene_short_name <- NULL # no visible binding

  # Yes, it's stupid we have cell ids both as a column and as the rownames.
  cell_group_df <- data.frame(row.names=row.names(SummarizedExperiment::colData(cds)),
                              cell_id=row.names(SummarizedExperiment::colData(cds)))

  # Set up the table that partitions the cells into groups.
  # Must be either a column in SummarizedExperiment::colData or one of "cluster" or "partition.
  # FIXME: Should check its not a column you can't really use for grouping
  # (i.e. a floating point value)
  if (group_cells_by == "cluster"){
    cell_group_df$cell_group <-
      tryCatch({monocle3::clusters(cds, reduction_method = reduction_method)},
               error = function(e) {NULL})
  } else if (group_cells_by == "partition") {
    cell_group_df$cell_group <-
      tryCatch({monocle3::partitions(cds, reduction_method = reduction_method)},
               error = function(e) {NULL})
  } else{
    cell_group_df$cell_group <- SummarizedExperiment::colData(cds)[,group_cells_by]
  }
  cell_group_df$cell_group <- as.character(cell_group_df$cell_group)

  if (verbose)
    message("Aggregating gene expression values for groups")
  # For each gene compute the fraction of cells expressing it within each group
  # in a matrix thats genes x cell groups

  cluster_binary_exprs = as.matrix(monocle3::aggregate_gene_expression(cds,
                                                             cell_group_df=cell_group_df,
                                                             norm_method="binary",
                                                             scale_agg_values=FALSE))

  cluster_mean_exprs = as.matrix(monocle3::aggregate_gene_expression(cds,
                                                           cell_group_df=cell_group_df,
                                                           norm_method="size_only",
                                                           scale_agg_values=FALSE))

  if (verbose)
    message("Computing Jensen-Shannon specificities")

  # Now compute a Jensen Shannon specificity score for each gene w.r.t each group
  cluster_spec_mat = specificity_matrix(cluster_mean_exprs, cores=cores)
  cluster_marker_score_mat = as.matrix(cluster_binary_exprs * cluster_spec_mat)

  if (verbose)
    message("Gathering score tables")
  cluster_marker_score_table = tibble::rownames_to_column(as.data.frame(cluster_marker_score_mat))
  cluster_marker_score_table = tidyr::gather(cluster_marker_score_table, "cell_group", "marker_score", -rowname)

  cluster_spec_table = tibble::rownames_to_column(as.data.frame(cluster_spec_mat))
  cluster_spec_table = tidyr::gather(cluster_spec_table, "cell_group", "specificity", -rowname)

  cluster_expr_table = tibble::rownames_to_column(as.data.frame(cluster_mean_exprs))
  cluster_expr_table = tidyr::gather(cluster_expr_table, "cell_group", "mean_expression", -rowname)

  cluster_fraction_expressing_table = tibble::rownames_to_column(as.data.frame(cluster_binary_exprs))
  cluster_fraction_expressing_table = tidyr::gather(cluster_fraction_expressing_table, "cell_group", "fraction_expressing", -rowname)

  # spec_model_df = data.frame(rowname=row.names(cluster_spec_score_mat),
  #                            num_expressing=Matrix::rowSums(SingleCellExperiment::counts(cds) > 0),
  #                            mean_exprs=Matrix::rowMeans(cluster_agg_exprs),
  #                            max_spec=rowMaxs(cluster_spec_score_mat))

  # # Now compute the expected max specificity as a function of how many cells express a given gene
  # # Genes that are expressed in few cells tend to have very high specificity, so we want to
  # # control for this trend when ranking genes by specificity later on
  # spec_model_df = spec_model_df %>% dplyr::mutate(quantile = dplyr::ntile(num_expressing, expression_bins))
  # spec_summary = spec_model_df %>% dplyr::group_by(quantile) %>% dplyr::summarize(log_spec_mean = mean(log(max_spec)), log_spec_sd = sd(log(max_spec)))
  # spec_model_df = dplyr::left_join(spec_model_df, spec_summary)
  #
  # # Compute the "specifity above expectation" for each gene w.r.t. each group:
  # cluster_spec_table = dplyr::left_join(cluster_spec_table, spec_model_df)
  # cluster_spec_table = cluster_spec_table %>% dplyr::mutate(log_spec=log(specificity),
  #                                                    pval_excess_spec = pnorm(log(specificity),log_spec_mean, log_spec_sd, lower.tail=FALSE))

  cluster_marker_score_table$specificity = cluster_spec_table$specificity
  cluster_marker_score_table$mean_expression = cluster_expr_table$mean_expression
  cluster_marker_score_table$fraction_expressing = cluster_fraction_expressing_table$fraction_expressing

  cluster_marker_score_table = cluster_marker_score_table %>%
    #filter(num_expressing > 10) %>%
    dplyr::group_by(cell_group) %>%
    dplyr::top_n(genes_to_test_per_group, marker_score)

  cell_group_df$cell_id <- as.character(cell_group_df$cell_id)
  cell_group_df$cell_group <- as.character(cell_group_df$cell_group)


  if (marker_sig_test){
    if (verbose)
      message("Running marker significance tests")
    # Temporarily disable OpenMP threading in functions to be run in parallel
    old_omp_num_threads = as.numeric(Sys.getenv("OMP_NUM_THREADS"))
    if (is.na(old_omp_num_threads)){
      old_omp_num_threads = 1
    }
    RhpcBLASctl::omp_set_num_threads(1)

    # Temporarily set the number of threads the BLAS library can use to be 1
    old_blas_num_threads = as.numeric(Sys.getenv("OPENBLAS_NUM_THREADS"))
    if (is.na(old_omp_num_threads)){
      old_blas_num_threads = 1
    }
    RhpcBLASctl::blas_set_num_threads(1)

    # Set up a balanced "reference" panel of cells from each group
    if (is.null(reference_cells) == FALSE){
      if(is.numeric(reference_cells)){
        num_ref_cells_per_group = reference_cells / length(unique(cell_group_df$cell_group))
        reference_cells = cell_group_df %>% dplyr::group_by(cell_group) %>%
          dplyr::sample_n(min(num_ref_cells_per_group, dplyr::n())) %>%
          dplyr::pull(cell_id)
        #reference_cells = sample(colnames(cds), reference_cells)
      } else {
        # TODO: check that reference cells is a list of valid cell ids.
      }
    }

    marker_test_res = tryCatch({pbmcapply::pbmcmapply(test_marker_for_cell_group,
                                                      cluster_marker_score_table$rowname,
                                                      cluster_marker_score_table$cell_group,
                                                      MoreArgs=list(cell_group_df, cds, reference_cells,
                                                                    speedglm.maxiter),
                                                      ignore.interactive = TRUE,
                                                      mc.cores=cores)},
                               finally = {
                                 RhpcBLASctl::omp_set_num_threads(old_omp_num_threads)
                                 RhpcBLASctl::blas_set_num_threads(old_blas_num_threads)
                               })

    #
    # Check for possible convergence failure or other problems. Issue: #383
    #
    if('warning' %in% names(marker_test_res)) {
      warning('test_marker_for_cell_group() caught warning: possible convergence failure.')
      warning(names(marker_test_res))
      marker_test_res <- marker_test_res[["value"]]
    }

    marker_test_res = t(marker_test_res)
    marker_test_res = as.matrix(marker_test_res)
    colnames(marker_test_res) = c("pseudo_R2", "lrtest_p_value")

    #marker_test_res = as.data.frame(marker_test_res)

    marker_test_res = dplyr::bind_cols(cluster_marker_score_table, as.data.frame(marker_test_res))
    marker_test_res$lrtest_q_value = stats::p.adjust(marker_test_res$lrtest_p_value,
                                                     method="bonferroni",
                                                     n=length(cluster_spec_mat))
    marker_test_res = marker_test_res %>% dplyr::select(rowname,
                                                        cell_group,
                                                        marker_score,
                                                        mean_expression,
                                                        fraction_expressing,
                                                        specificity,
                                                        pseudo_R2,
                                                        lrtest_p_value,
                                                        lrtest_q_value)

    marker_test_res = marker_test_res %>% dplyr::rename(gene_id=rowname, marker_test_p_value=lrtest_p_value,  marker_test_q_value=lrtest_q_value)
    marker_test_res$pseudo_R2 = unlist(marker_test_res$pseudo_R2)
    marker_test_res$marker_test_p_value = unlist(marker_test_res$marker_test_p_value)
    if ("gene_short_name" %in% colnames(SummarizedExperiment::rowData(cds))){
      marker_test_res = SummarizedExperiment::rowData(cds) %>%
        as.data.frame %>%
        tibble::rownames_to_column() %>%
        dplyr::select(rowname, gene_short_name) %>%
        dplyr::inner_join(marker_test_res, by=c("rowname"="gene_id"))
      marker_test_res = marker_test_res %>% dplyr::rename(gene_id=rowname)
    }
  } else {
    marker_test_res = cluster_marker_score_table
    marker_test_res = marker_test_res %>% dplyr::rename(gene_id=rowname)
  }

  if (verbose)
    message("Done")

  return(marker_test_res)
}


# Calculate the probability vector
makeprobsvec <- function(p) {
  phat <- p/sum(p)
  phat[is.na(phat)] = 0
  phat
}

# Calculate the probability matrix for a relative abundance matrix
makeprobs <- function(a) {
  colSums<-apply(a,2,sum)
  b <- Matrix::t(Matrix::t(a)/colSums)
  b[is.na(b)] = 0
  b
}

# Calculate the Shannon entropy based on the probability vector
shannon.entropy <- function(p) {
  if (min(p) < 0 || sum(p) <=0)
    return(Inf)
  p.norm <- p[p>0]/sum(p)
  -sum(log2(p.norm)*p.norm)
}

# Calculate the Jessen-Shannon distance for two probability distribution
JSdistVec <- function (p, q)
{
  JSdiv <- shannon.entropy((p + q)/2) - (shannon.entropy(p) +
                                           shannon.entropy(q)) * 0.5
  JSdiv[is.infinite(JSdiv)] <- 1
  JSdiv[JSdiv < 0] <- 0
  JSdist <- sqrt(JSdiv)
  JSdist
}

specificity_matrix <- function(agg_expr_matrix, cores=1){
  if(ncol(agg_expr_matrix) < 1) warning('bad loop: ncol(agg_expr_matrix) < 1')
  specificity_mat <-
    pbmcapply::pbmclapply(row.names(agg_expr_matrix),
                          FUN = function(x) {
                            agg_exprs = as.numeric(agg_expr_matrix[x,])
                            agg_exprs = makeprobsvec(agg_exprs)
                            perfect_spec_matrix = diag(ncol(agg_expr_matrix))
                            sapply(1:ncol(agg_expr_matrix), function(col_idx) {
                              1 - JSdistVec(agg_exprs,
                                            perfect_spec_matrix[,col_idx])
                            })
                          }, mc.cores=cores,
                          ignore.interactive = TRUE)
  specificity_mat = do.call(rbind, specificity_mat)
  colnames(specificity_mat) = colnames(agg_expr_matrix)
  row.names(specificity_mat) = row.names(agg_expr_matrix)
  return(specificity_mat)
  #
}

enrichment_matrix <- function(agg_expr_matrix, cores=1){
  if(ncol(agg_expr_matrix) < 1) warning('bad loop: ncol(agg_expr_matrix) < 1')
  specificity_mat = pbmcapply::pbmclapply(row.names(agg_expr_matrix),
                                          FUN = function(x)
                                          {
                                            agg_exprs = as.numeric(agg_expr_matrix[x,])
                                            agg_exprs = makeprobsvec(agg_exprs)
                                            perfect_spec_matrix = diag(ncol(agg_expr_matrix))
                                            sapply(1:ncol(agg_expr_matrix), function(col_idx) {
                                              1 - JSdistVec(agg_exprs, perfect_spec_matrix[,col_idx])
                                            }
                                            )
                                          }, mc.cores=cores,
                                          ignore.interactive = TRUE)
  specificity_mat = do.call(rbind, specificity_mat)
  colnames(specificity_mat) = colnames(agg_expr_matrix)
  row.names(specificity_mat) = row.names(agg_expr_matrix)
  return(specificity_mat)
  #
}

test_marker_for_cell_group = function(gene_id, cell_group, cell_group_df, cds,
                                      reference_cells=NULL, speedglm.maxiter=25){
  #print(gene_id)
  #print(cell_group)
  #print (length(reference_cells))
  results <- tryCatch({
    f_expression <-
      log(as.numeric(SingleCellExperiment::counts(cds)[gene_id,]) / size_factors(cds) + 0.1)
    #print(sum(SingleCellExperiment::counts(cds)[gene_id,] > 0))
    is_member <-
      as.character(cell_group_df[colnames(cds),2]) == as.character(cell_group)
    names(is_member) = names(f_expression) = colnames(cds)
    is_member[is.na(is_member)] = FALSE
    is_member[is.null(is_member)] = FALSE

    if (is.null(reference_cells) == FALSE){
      # Exclude cells that aren't in either the cell_group or the
      # reference_panel
      f_expression <- f_expression[is_member | names(f_expression) %in%
                                     reference_cells]
      is_member <- is_member[is_member | names(is_member) %in% reference_cells]
    }

    if (sum(is.na(f_expression)) > 0 || sum(is.na(is_member)) > 0){
      stop("Expression and group membership can't be NA")
    }
    model <- speedglm::speedglm(is_member ~ f_expression,
                                acc=1e-3, model=FALSE,
                                y=FALSE,
                                verbose=TRUE,
                                family=stats::binomial(),
                                maxit=speedglm.maxiter)
    assertthat::assert_that(model$convergence == TRUE, msg=paste0('speedglm model failed to converge in ',speedglm.maxiter, ' iterations.'))
    null_model <- speedglm::speedglm(is_member ~ 1,
                                     acc=1e-3, model=FALSE,
                                     y=FALSE,
                                     verbose=TRUE,
                                     family=stats::binomial(),
                                     maxit=speedglm.maxiter)
    assertthat::assert_that(model$convergence == TRUE, msg=paste0('speedglm null model failed to converge in ',speedglm.maxiter, ' iterations.'))
    lr.stat <- lmtest::lrtest(null_model, model)
    #print (summary(model))
    # #print(summary(null_model))
    # #print (lr.stat)
    #print (str(lr.stat))
    n=ncol(cds)
    pseudo_R2 <-
      (1-exp(-as.numeric(lr.stat$Chisq[2])/n)) /
      (1-exp(2*as.numeric(stats::logLik(null_model)/n)))
    LR_test_pval = lr.stat$`Pr(>Chisq)`[2]
    # model_summary = summary(model)
    # #print(model_summary)
    # #pval = as.numeric(as.character(model_summary$coefficients[2,4]))
    # pseudo_R2
    #pval
    return (list(pseudo_R2, LR_test_pval))
  }, error = function(e) { return(list(0.0, 1.0)) })

  #print(pval)
  return(results)
}

#' Generate a Garnett marker file from top_markers output.
#'
#' @param marker_test_res Tibble of top markers, output of
#'   \code{\link{top_markers}}.
#' @param file Path to the marker file to be generated. Default is
#'   "./marker_file.txt".
#' @param max_genes_per_group Numeric, the maximum number of genes to output
#'   per cell type entry. Default is 10.
#' @param remove_duplicate_genes Logical indicating whether marker genes that
#'   mark multiple cell groups should be excluded. Default is FALSE. When
#'   FALSE, a message will be emitted when duplicates are present.
#'
#' @return None, marker file is written to \code{file} parameter location.
#' @export
#'
generate_garnett_marker_file <- function(marker_test_res,
                                         file = "./marker_file.txt",
                                         max_genes_per_group = 10,
                                         remove_duplicate_genes = FALSE) {
  group_name <- marker_score <- NULL # no visible binding
  marker_test_res <- as.data.frame(marker_test_res)
  if(is.null(marker_test_res$group_name)) {
    marker_test_res$group_name <- paste("Cell type", marker_test_res$cell_group)
  }
  group_list <- unique(marker_test_res$group_name)

  good_markers <- marker_test_res %>% dplyr::group_by(group_name) %>%
    dplyr::top_n(max_genes_per_group, marker_score)

  dups <- good_markers$gene_id[duplicated(good_markers$gene_id)]

  if ("gene_short_name" %in% colnames(good_markers)) {
    dups_gsn <- good_markers$gene_short_name[duplicated(good_markers$gene_short_name)]
  }


  if(remove_duplicate_genes) {
    good_markers <- good_markers[!good_markers$gene_id %in% dups,]
  } else {
    if (length(dups) > 0) {
      if("gene_short_name" %in% colnames(good_markers)) {
        message("The following marker genes mark multiple cell groups. ",
                "Prior to using Garnett, we recommend either excluding ",
                "these genes using remove_duplicate_genes = TRUE, or ",
                "modifying your marker file to make the cell types with ",
                "the shared marker subtypes in a hierarchy. ",
                paste(dups_gsn, collapse = ", "))
      } else {
        message("The following marker genes mark multiple cell groups. ",
                "Prior to using Garnett, we recommend either excluding ",
                "these genes using remove_duplicate_genes = TRUE, or ",
                "modifying your marker file to make the cell types with ",
                "the shared marker subtypes in a hierarchy. ",
                paste(dups, collapse = ", "))
      }
    }
  }

  output <- list()

  for (group in group_list) {
    if (sum(good_markers$group_name == group) == 0) {
      message(group, " did not have any markers above the q-value ",
              "threshold. It will be skipped.")
      next
    }

    good_name <- gsub("\\(|\\)|:|>|,|#", ".", group)
    if (good_name != group) {
      warning("Group name contained an illegal character for a Garnett ",
              "marker file. ", group, " will be substituted for ",
              good_name, ".")
    }

    sub <- good_markers[good_markers$group_name == group,]
    if (nrow(sub) > max_genes_per_group) {
      if(max_genes_per_group < 1) warning('bad loop: max_genes_per_group < 1')
      sub <- sub[order(sub$marker_test_q_value),][1:max_genes_per_group,]
    }
    if ("gene_short_name" %in% colnames(sub)){
      entry <- paste0("> ", good_name, "\n", "expressed: ",
                      paste(sub$gene_short_name, collapse = ", "), "\n")
    } else {
      entry <- paste0("> ", good_name, "\n", "expressed: ",
                      paste(sub$gene_id, collapse = ", "), "\n")
    }

    output <- append(output, entry)
  }

  all <- paste(output, collapse = "\n")

  write(all, file=file)
  message("Garnett marker file written to ", file)
}
```

# Run info {-}

```{r curr_params, echo = F}
curr_params %>%
  flatten_list() %>%
  {tibble::tibble(
    param = names(.), 
    value = purrr::map(., paste, collapse = ", ") %>% unlist() %>% substr(1, 100))
  } %>%
  knitr::kable()
```

# Output info {-}

Output files will be saved to **`r curr_params$output$dir`**.

Figures will be saved to **`r curr_params$output$dir`qc_report_files/figure-html/**.

Intermediate objects will be cached to **`r curr_params$output$dir`qc_report_cache/html/**.

# `Seurat` pre-processing

## Detect snRNA-seq platform

``` {r detect_platform}
input_files <- list.files(curr_params$input$dir)
platform_files <- list(
  "Parse Biosciences" = c("DGE.mtx", "all_genes.csv", "cell_metadata.csv"),
  "10X Genomics" = c("matrix.mtx.gz", "features.tsv.gz", "genes.tsv.gz", "barcodes.tsv.gz")
)
platform <-
  platform_files %>%
  purrr::map(~ (.x %>% intersect(input_files) %>% length) == 3) %>%
  unlist() %>%
  {platform_files[.]} %>%
  names
```

The input data comes from the *`r platform`* platform and includes `r paste(input_files, sep = ", ")` files.

## Load data and apply gene-level filter

Load input Seurat object with no cut-offs, remove any missing samples, add sample metadata, add summary statistics, and optionally subset samples. 

Then, filter nuclei per gene, based on the `n_nuclei_per_gene$min` argument. Genes that are present in very few nuclei of the dataset (e.g., <3) are uninformative and unlikely to help in differentiating between groups of nuclei. In general, most genes removed by this filtering step will be those not detected in any nuclei, which will help to trim the size of the object. Additionally, genes are removed that have entries in all nuclei but whose entries are all zero. 

The minimum number of nuclei per gene is set to **`r curr_params$filter$n_feature_RNA$min`**.

``` {r load_data, warning = F}
seu <- xfun::cache_rds({
  
  if (platform == "10X Genomics") {
    
    # read in expression matrix
    dge_mat <- Seurat::Read10X(curr_params$input$dir)
    
    # create seurat object
    seu <- Seurat::CreateSeuratObject(counts = dge_mat)
    
  } else if (platform == "Parse Biosciences") {
    
    # read in expression matrix
    dge_mat <- Seurat::ReadParseBio(curr_params$input$dir)
    
    # read in cell metadata
    cell_metadata <-
      curr_params$input$dir %>%
      paste0("/cell_metadata.csv") %>%
      readr::read_csv(show_col_types = F) %>%
      tibble::column_to_rownames("bc_wells") %>%
      as.data.frame()
    
    # create seurat object
    seu <-
      dge_mat %>%
      Seurat::CreateSeuratObject(
        names.field = 0,
        meta.data = cell_metadata,
        min_genes = curr_params$filter$n_feature_RNA$min,
        min_cells = curr_params$filter$n_nuclei_per_gene$min
      )
    
  }
  
  # conditionally add sample metadata
  if (curr_params$input$sample_metadata_file != F) {

    # read in sample metadata and add to misc slot
    seu@misc$sample_metadata <-
      readr::read_tsv(curr_params$input$sample_metadata_file, show_col_types = F)

    # add to Seurat object meta.data
    seu@meta.data <-
      dplyr::left_join(
        seu@meta.data,
        seu@misc$sample_metadata %>%
          dplyr::select(sample, dplyr::any_of(groupings)),
        by = "sample")
    rownames(seu@meta.data) <- colnames(seu)

  }
  
  # remove genes with no counts
  counts <- as.matrix(seu@assays$RNA@counts)
  seu <- subset(seu, features = rownames(counts[rowSums(counts) > 0, ]))
  
  # return
  seu
  
}, file = "seu.rds")
```

```{r summary_statistics_plot, class.source = 'fold-hide', message = F, warning = F, fig.height = 15, fig.cap = "Plot of summary statistics by sample"}
p <-
  seu@meta.data %>%
  tibble::as_tibble() %>%
  dplyr::group_by(sample) %>%
  dplyr::summarise(
    n_nuclei = dplyr::n(),
    median_nCount_RNA = median(nCount_RNA),
    median_nFeature_RNA = median(nFeature_RNA)
  ) %>%
  tidyr::pivot_longer(-sample, names_to = "statistic") %>%
  ggplot2::ggplot(ggplot2::aes(x = sample, y = value)) +
  ggplot2::geom_col() +
  ggplot2::facet_grid(statistic ~ ., scales = "free") +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90)) +
  ggplot2::theme_classic()
p
```

## Filter nuclei

Nucleus-level filtering is performed, applying cut-offs to `nCount_RNA`, `nFeature_RNA`, `percent_mito`, and `doublet` values of each nucleus.

* Unusually high transcript / gene counts indicate multiplets.
* Unusually low transcript / gene counts indicate barcoding of nuclei with damaged membranes, which are likely low-quality and uninformative.
* A high percentage of mitochondrial genes indicates death, loss of cytoplasmic RNA, or increased apoptosis. Differences in the application of this filter between scRNA-seq and snRNA-seq are not well covered in literature. Intuitively, snRNA-seq data should be treated with a more stringent cut-off, as it should only include nuclear RNA. However, the renal cortex has an exceptionally high level of mitochondrial activity, so this cut-off has been relaxed somewhat for the VHL dataset. 
* Nuclei with a high nFeature_RNA:nCount_RNA ratio could be from dying cells.

### Annotate doublets

Putative doublets are identified using the `scDblFinder` package. This package creates artificial doublets by grouping together random pairs of nuclei within each sample to create pseudo-doublets, and then identifying nuclei that cluster near to these pseudo-doublets during dimensionality reduction. This co-clustering indicates that they carry a doublet signature. 

The `remove_doublets` argument is set to **`r  curr_params$filter$doublets`**.
`r if (curr_params$filter$doublets == T) {"Doublets will be identified and removed." }`
`r if (curr_params$filter$doublets == F) {"A dummy column will be introduced, treating all nuclei as though they are singlets."}`

```{r identify_doublets, eval =  curr_params$filter$doublets == T, include =  curr_params$filter$doublets == T}
seu <- xfun::cache_rds({
  seu_dbl <- scDblFinder::scDblFinder(
    Seurat::as.SingleCellExperiment(seu), samples = "sample"
  )
  seu$doublet <- as.numeric(seu_dbl$scDblFinder.class == "doublet")
  seu
}, file = "seu_doublets.rds")
```

```{r ignore_doublets, eval =  curr_params$filter$doublets == F, include =  curr_params$filter$doublets == F}
seu$doublet <- 0
```

```{r detected_doublets, eval =  curr_params$filter$doublets == T, class.source = 'fold-hide', fig.cap = "Plot of detected doublets in each sample (using scDblFinder)"}
p_dat <- 
  seu@meta.data[,c("sample", "doublet", "nCount_RNA", "nFeature_RNA")] %>% 
  dplyr::group_by(sample) %>%
  dplyr::mutate(is_doublet = doublet == 1,
                n_doublets = sum(doublet),
                n_singlets = sum(!doublet))
p1 <- 
  p_dat %>% 
  tidyr::pivot_longer(c("n_doublets", "n_singlets"), values_to = "count") %>%
  ggplot2::ggplot(ggplot2::aes(y = count, x = sample, fill = name)) +
  ggplot2::geom_col() +
  ggplot2::scale_fill_manual(
    name = "",
    values = c(n_singlets = dittoSeq::dittoColors()[1],
               n_doublets = dittoSeq::dittoColors()[2])) +
  ggplot2::theme(axis.text = ggplot2::element_text(size = 6),
                 axis.text.x = ggplot2::element_text(angle = 90),
                 legend.position = "none") 
p2 <-
  p_dat %>%
  tidyr::pivot_longer(c("nCount_RNA", "nFeature_RNA"), values_to = "count") %>%
  ggplot2::ggplot(ggplot2::aes(y = count, x = sample)) +
  ggplot2::geom_jitter(data = . %>% dplyr::filter(!is_doublet), height = 0,
                       ggplot2::aes(colour = "singlet"), size = 0.2) +
  ggplot2::geom_jitter(data = . %>% dplyr::filter(is_doublet), height = 0,
                       ggplot2::aes(colour = "doublet"), size = 0.2) +
  ggplot2::facet_wrap(~ name, scales = "free") +
  ggplot2::scale_colour_manual(
    name = "",
    values = c(singlet = dittoSeq::dittoColors()[1],
               doublet = dittoSeq::dittoColors()[2])) +
  ggplot2::theme(axis.text = ggplot2::element_text(size = 7),
                 axis.text.x = ggplot2::element_text(angle = 90),
                 axis.title.y = ggplot2::element_blank(),
                 legend.position = "bottom") +
  patchwork::plot_layout(ncol = 2, widths = c(1, 2))
p1 + p2
```

### Annotate transcript types

Proportions of transcript types of interest (mitochondrial, ribosomal, globin) are annotated. Mitochondrial genes are particularly of interest for filtering, but it is good practice to inspect the others as well.

* `percent_mito` - A high percentage of mitochondrial genes indicates death, loss of cytoplasmic RNA, or increased apoptosis.
* `percent_ribo` - mRNA that code for ribosomal proteins. They do not point to specific issues, but it can be good to have a look at their relative abundance. They can have biological relevance.
* `percent_globin` - Globin genes are very abundant in erythrocytes. Depending on your application, you can expect ‘contamination’ of erythrocytes and select against it.

```{r annotate_transcript_types}
transcript_types <-
  list(mito = "^MT-",
       ribo = "^RP[SL]",
       globin = "^HB[^(P)]") 

transcript_types %>%
  names %>%
  purrr::map(function(tt) {
    # use `<<` for global assignment
    seu <<-
      seu %>%
      Seurat::PercentageFeatureSet(
        pattern = transcript_types[[tt]],
        col.name = paste0("percent_", tt))
  })
```

### Define nucleus-level filters

The `filter.do_filtering` argument is set to **`r curr_params$filter$do_filtering`**.
`r if (curr_params$filter$do_filtering == T) {"Filters will be applied." }`
`r if (curr_params$filter$do_filtering == F) {"Dummy filters will be set (min = -Inf, max = Inf) and all nuclei will be retained." }`

```{r get_filters}
# seu <- get_filters(
#   seu,
#   curr_params$filter$do_filtering,
#   curr_params$filter$doublet,
#   curr_params$filter$nCount_RNA$min,
#   curr_params$filter$nCount_RNA$max,
#   curr_params$filter$nFeature_RNA$min,
#   curr_params$filter$nFeature_RNA$max,
#   curr_params$filter$percent_mito$max
# )

# get filters (user-provided or median + n_mads * MAD)
if(curr_params$filter$do_filtering == T) {
  filters <- curr_params$filter[c("nCount_RNA", "nFeature_RNA", "percent_mito")]
  filters$doublet <- list(min = 0, max = ifelse(curr_params$filter$doublets == T, 0, 1))
} else {
  filters <- list(
    "doublet" = list(min = -Inf, max = Inf),
    "nCount_RNA" = list(min = -Inf, max = Inf),
    "nFeature_RNA" = list(min = -Inf, max = Inf),
    "percent_mito" = list(min = -Inf, max = Inf)
  )
}

# add to seu
seu@misc$filters <- filters

# create misc table of pass/fail nuclei, with fail criteria
seu@misc$nucleus_filtering <-
  seu@meta.data[, colnames(seu@meta.data) %in% c("sample", names(filters))] %>%
  dplyr::as_tibble(rownames = "nucleus") %>%
  tidyr::pivot_longer(-c(nucleus, sample)) %>%
  dplyr::rowwise() %>%
  dplyr::mutate(
    fail_criteria = ifelse(
      value < filters[[name]]$min, paste0("min_", name),
      ifelse(
        value > filters[[name]]$max, paste0("max_", name),
        NA
      )
    ),
    pass = is.na(fail_criteria)) %>%
  dplyr::group_by(nucleus, sample) %>%
  dplyr::mutate(pass = all(pass),
                fail_criteria = fail_criteria %>% unique %>% na.omit %>% paste(collapse = ",") %>% dplyr::na_if(., "")) %>%
  tidyr::pivot_wider(id_cols = c("nucleus", "sample", "pass", "fail_criteria")) %>%
  dplyr::ungroup()
```

#### Fail criteria statistics

```{r calc_n_retained, include = F}
n_retained <- nrow(dplyr::filter(seu@misc$nucleus_filtering, pass))
n_total <- nrow(seu@misc$nucleus_filtering)
```

`r paste0("**", round((n_retained / n_total) * 100, 1), "%** (", n_retained, " / ", n_total, ") of nuclei retained.")`

```{r fail_criteria_stats, echo = F, message = F}
options(knitr.kable.NA = '')
dplyr::left_join(
  seu@misc$filters %>% dplyr::bind_rows(.id = "fail_criteria") %>%
    tidyr::pivot_longer(c("min", "max")) %>%
    dplyr::transmute(fail_criteria = paste(name, fail_criteria, sep = "_"), value),
  seu@misc$nucleus_filtering %>%
  dplyr::filter(!pass) %>%
  tidyr::separate_longer_delim(fail_criteria, ",") %>%
  dplyr::count(fail_criteria) %>%
  dplyr::mutate(`%` = round((n / n_total) * 100, 2)),
  by = "fail_criteria"
) %>%
  dplyr::arrange(-n) %>%
  dplyr::bind_rows(dplyr::tibble(fail_criteria = "total:", 
                                 n = n_total - n_retained,
                                 `%` = round((n / n_total) * 100, 2))) %>%
  knitr::kable(align = c("c", "c", "c", "c", "c"))
```

```{r fail_criteria_plot, class.source = 'fold-hide', fig.cap = "Plot of the proportions of filtered nuclei and their fail criteria for each sample"}
p <- 
  seu@misc$nucleus_filtering %>%
  dplyr::group_by(sample) %>%
  dplyr::count(fail_criteria) %>%
  ggplot2::ggplot(ggplot2::aes(x = sample, y = n, fill = fail_criteria)) +
  ggplot2::geom_col() +
  ggplot2::theme_minimal() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90)) +
  ggplot2::scale_fill_manual(values = dittoSeq::dittoColors(), na.value = "grey")
p
```

```{r filters_vs_distribution_vln_plot, class.source = 'fold-hide', message = F, warning = F, fig.height = 10, fig.width = 10, fig.cap = "Plot of filters vs feature distribution"}
p <- 
    seu@misc$nucleus_filtering %>%
    tidyr::pivot_longer(names(seu@misc$filters)) %>%
    dplyr::left_join(seu@misc$filters %>%
                       purrr::map(~dplyr::as_tibble(.)) %>%
                       dplyr::bind_rows(.id = "name"),
                     by = "name") %>%
    dplyr::group_by(fail_criteria) %>%
    ggplot2::ggplot(ggplot2::aes(x = sample, y = value)) +
    ggplot2::geom_jitter(ggplot2::aes(colour = pass),
                         height = 0, alpha = 0.5, size = 0.6) +
    ggplot2::geom_violin(ggplot2::aes(fill = sample), alpha = 0.8,
                         draw_quantiles = 0.5) +
    ggplot2::geom_hline(ggplot2::aes(yintercept = min), colour = "red") +
    ggplot2::geom_hline(ggplot2::aes(yintercept = max), colour = "red") +
    ggplot2::facet_wrap(~ name, scales = "free") +
    ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90)) +
    ditto_colours +
    ggplot2::scale_colour_manual(values = c("darkgrey", "black"))
p
```

```{r log_filters_vs_distribution_vln_plot, class.source = 'fold-hide', message = F, warning = F, fig.height = 10, fig.width = 10, fig.cap = "Plot of filters vs feature distribution (log scale)"}
p + ggplot2::scale_y_log10()
```

```{r nucleus_scatter_with_filters_plot, class.source = 'fold-hide', fig.width = 10, fig.cap = "Plot of filters against distribution of nuclei"}
plot_nucleus_scatter_with_filters <- function(seu, x, y, ditto_colours, log_x = F, log_y = F) {
  dat <- seu@misc$nucleus_filtering %>%
    dplyr::mutate(x_var = get(x) %>% { if(log_x) log(.) else . },
                  y_var = get(y) %>% { if(log_y) log(.) else . },
                  title = paste0(x, " n fail = ", sum(grepl(x, fail_criteria)), " / ", dplyr::n(), "\n",
                                 y, " n fail = ", sum(grepl(y, fail_criteria)), " / ", dplyr::n()))
  dat %>%
    ggplot2::ggplot(ggplot2::aes(x = x_var, y = y_var, colour = fail_criteria)) +
    ggplot2::geom_vline(xintercept = unlist(seu@misc$filters[[x]]) %>% { if(log_x) log(.) else . },
                        colour = "red") +
    ggplot2::geom_hline(yintercept = unlist(seu@misc$filters[[y]]) %>% { if(log_y) log(.) else . },
                        colour = "red") +
    ggplot2::geom_point(size = 0.5, alpha = 0.7) +
    ggplot2::geom_point(data = . %>% dplyr::filter(pass == F), size = 0.5, alpha = 0.7) +
    ggplot2::labs(title = unique(dat$title),
                  x = paste0(ifelse(log_x, "log ", ""), x),
                  y = paste0(ifelse(log_y, "log ", ""), y)) +
    ditto_colours +
    ggplot2::theme(plot.title = ggplot2::element_text(size = 10))

}
p <- 
    plot_nucleus_scatter_with_filters(seu, "nCount_RNA", "percent_mito", ditto_colours) +
    ggplot2::theme(legend.position = "none") +
    plot_nucleus_scatter_with_filters(seu, "nCount_RNA", "nFeature_RNA", ditto_colours)
p
```

```{r log_nucleus_scatter_with_filters_plot, class.source = 'fold-hide', fig.width = 10, fig.cap = "Plot of filters against distribution of nuclei (log scale)"}
p + ggplot2::scale_x_log10() + ggplot2::scale_y_log10()
```

### `r if (curr_params$filter$do_filtering) {"Apply nucleus-level filters"}`

```{r filter_nuclei, eval = curr_params$filter$do_filtering == T, include = curr_params$filter$do_filtering == T}
seu <- xfun::cache_rds({
  # remove failing nuclei 
  seu <- subset(seu, cells = unique(dplyr::filter(seu@misc$nucleus_filtering, pass)$nucleus))
  # remove genes with no counts
  counts <- as.matrix(seu@assays$RNA@counts)
  seu <- subset(seu, features = rownames(counts[rowSums(counts) > 0, ]))
  seu
}, file = "seu_filtered.rds")
```

# `Seurat` processing

## Find highest variable genes

Find the most highly variable genes (HVGs) across samples.

```{r hvg, message = F, warning = F}
seu <- xfun::cache_rds({
  Seurat::FindVariableFeatures(seu)
}, file = "seu_hvg.rds")
```

```{r hvg_scatter_plot, class.source = 'fold-hide', warning = F, message = F, fig.cap = "Plot of HVGs"}
p <- 
  Seurat::LabelPoints(
    plot = Seurat::VariableFeaturePlot(seu, raster = F),
    points = head(Seurat::VariableFeatures(seu), 10),
    repel = T
  ) +
    ggplot2::theme(legend.position = "top")
p
```

```{r hvg_heatmap_plot, class.source = 'fold-hide', fig.cap = "Heatmap of top 20 HVGs"}
p <- 
  dittoSeq::dittoHeatmap(
    seu,
    genes = head(Seurat::VariableFeatures(seu), 20),
    annot.by = "sample",
    scaled.to.max = T
  )
p
```

## Normalisation and scaling

`Seurat::SCTransform()` performs normalisation and scaling of the data. 

Variables to regress out of the `SCTransform` residuals  (set by argument `vars_to_regress`) are **`r paste(curr_params$vars_to_regress, collapse = "** and **")`**. These variables are prevented from contributing much to the PCA during dimensionality reduction, and thus confounding the analysis, as they are considered to be a source of irrelevant variation.

```{r sctransform, results = F, warning = F}
seu <- xfun::cache_rds({
  Seurat::SCTransform(seu, vars.to.regress = curr_params$vars_to_regress)
}, file = "seu_transformed.rds")
```

## Annotate cell cycle phases

Nuclei are assigned a score representing the likelihood that they are in each phase of the cell cycle. This is based on the expression levels of genes associated with each phase. The highest probability phase of each nucleus is stored in a new "Phase" column of the metadata.

```{r annotate_cell_cycle_phases}
seu <- Seurat::CellCycleScoring(
  seu,
  s.features = Seurat::cc.genes$s.genes,
  g2m.features = Seurat::cc.genes$g2m.genes
)
```

## Defining dimensionality

Dimensionality reduction is performed to isolate the strongest signals and to remove background noise. First, we must perform linear dimensionality reduction and then decide the dimensionality of the data.

### Linear dimensionality reduction (PCA)

```{r run_pca}
seu <- xfun::cache_rds({
  Seurat::RunPCA(seu)
}, file = "seu_pca.rds")
```

```{r pc1_vs_pc2_plot, class.source = 'fold-hide', fig.cap = "Plot of PC1 vs PC2, coloured by sample"}
p <- 
  dittoSeq::dittoDimPlot(seu, "sample", reduction.use = "pca", raster = F) 
p
```

```{r pca_dim, class.source = 'fold-hide', fig.cap = "Plot of cells and genes sorted by their PC scores for PC1 and PC2"}
p <- 
  Seurat::DimHeatmap(
    seu,
    dims = 1:2,
    cells = 500,
    balanced = TRUE,
    raster = F,
    fast = F
  )
p
```

```{r top_genes_pc1_pc2, class.source = 'fold-hide', fig.cap = "Plot of top genes associated with PC1 and PC2"}
p <- 
  Seurat::VizDimLoadings(seu, dims = 1:2, reduction = "pca")
p
```

### Dimensionality of the data

The dimensionality of the dataset to use for downstream analysis must be decided. This is the number of principal components believed to capture the majority of true biological signal in the dataset. Using too many PCs usually does not affect the result too much, while including too few PCs can be detrimental, as meaningful biological variation may be lost. Therefore, it is better to be overgenerous when defining the dimensionality of the data than to underestimate it.

The dimensionality can be decided by consulting the elbow plot (see below) and manually picking a value (set by argument `n_dims`) or it can be calculated using the `intrinsicDimension` package. 

```{r usr_n_dims_msg, echo = F, eval = curr_params$n_dims != F, results = 'asis'}
cat("### Manually\n")
cat(paste0("Dimensionality of the data is **", curr_params$n_dims, "** (set by argument n_dims).\n"))
```

```{r usr_n_dims, include = curr_params$n_dims != F, eval = curr_params$n_dims != F}
final_n_dims <- curr_params$n_dims
```

#### intrinsicDimension

```{r iD_msg, echo = F, eval = curr_params$n_dims == F, results = 'asis'}
cat("No value was provided for `n_dims`, so dimensionality will be calculated using the intrinsicDimensions package.\n")
```

The dimensionality of the dataset can be estimated using the `intrinsicDimension` package. This bypasses the manual approach of reading the elbow plot, but can sometimes be a bit too conservative. Therefore, if `n_dims` is not defined by the user, the dimensionality is set as double the `intrinsicDimension` estimate. 

```{r iD_n_dims}
n_dims_iD <- xfun::cache_rds({
  intrinsicDimension::maxLikGlobalDimEst(
    seu@reductions$pca@cell.embeddings,
    k = 10
  )$dim.est %>% round(0)
}, file = "n_dims_iD.rds")
generous_n_dims <- round(2 * n_dims_iD, 0)
```

```{r iD_n_dims_set, include = curr_params$n_dims == FALSE, eval = curr_params$n_dims == FALSE}
final_n_dims <- generous_n_dims
```

Dimensionality set to **`r final_n_dims`**.

```{r elbow_plot, class.source = 'fold-hide', fig.cap = "Elbow plot"}
p <-
  Seurat::ElbowPlot(seu, ndims = max(50, final_n_dims)) &
  ggplot2::geom_vline(
    ggplot2::aes(xintercept = final_n_dims,
                 colour = ifelse(curr_params$n_dims == F, "chosen n_dims", "user-provided n_dims")),
                 linewidth = 1) &
  ggplot2::geom_vline(ggplot2::aes(xintercept = n_dims_iD, colour = "intrisicDimensions"),
                      linewidth = 1, linetype = "dashed") &
  ggplot2::geom_vline(ggplot2::aes(xintercept = generous_n_dims, colour = "generous (iD x 2)"),
                      linewidth = 1, linetype = "dashed") &
  ggplot2::scale_colour_manual(
    name = "n_dims cut-off",
    values = c(`chosen n_dims` = dittoSeq::dittoColors()[[1]],
               `user-provided n_dims` = dittoSeq::dittoColors()[[1]],
               intrisicDimensions = dittoSeq::dittoColors()[[2]],
               `generous (iD x 2)` = dittoSeq::dittoColors()[[3]])) &
  ggplot2::theme(legend.position = c(0.7, 0.9),
                 legend.background = ggplot2::element_rect(fill = "white"))
p
```

# `monocle3` processing

The `monocle3` package provides alternative processing approaches for single cell data.   

First, we must convert the data to monocle3's proprietary `cell_data_set` object and fill in gene names and cell cycle scores.

```{r seu_to_cds, results = F, warning = F}
cds <- xfun::cache_rds({
  cds <- SeuratWrappers::as.cell_data_set(seu, assay = "RNA")
  SummarizedExperiment::rowData(cds)$gene_short_name <- 
    row.names(SummarizedExperiment::rowData(cds))
  SummarizedExperiment::colData(cds)$Phase <-
    seu$Phase
  cds
}, file = "cds.rds")
```

## Re-pre-processing and linear dimensionality reduction

First, we perform normalisation and principal component analysis again. We use the previously decided dimensionality of the data (**`r final_n_dims`**) from the `Seurat` section.

```{r preprocessing}
cds <- xfun::cache_rds({
  monocle3::preprocess_cds(cds, num_dim = final_n_dims)
}, file = "cds_lin_dim_red.rds")
```

## Non-linear dimensionality reduction 

Next, we perform non-linear dimensionality reduction of the data (UMAP). 

```{r nonlin_dim_redu, results = F, warning = F, message = F}
cds <- xfun::cache_rds({
  monocle3::reduce_dimension(cds, preprocess_method = "PCA")
}, file = "cds_nonlin_dim_red.rds")
```

Plotting all different groupings of the data onto the reductions can help reveal batch effects. If certain experimental variables seem to strongly influence the grouping of cells, these effects can be removed. If so, use the `align_cds` step from [this tutorial](https://cole-trapnell-lab.github.io/monocle3/docs/clustering/#reduce-dimension).

```{r avail_groupings, include = F}
# amend groupings
avail_groupings <- groupings[groupings %in% colnames(SummarizedExperiment::colData(cds))]
# reductions
redus <- c("PCA", "UMAP")
```

```{r groupings_vs_reductions, class.source = 'fold-hide', fig.dim = get_fig_dims(n_plots = (length(redus) + 1) * length(avail_groupings), n_cols = 3), fig.cap = "Plot of all monocle3 dimensionality reductions, coloured by groupings"}
# plot all groupings vs all reductions
groupings_vs_reductions <- list()
purrr::walk(avail_groupings, function(grouping) {
  purrr::walk(redus, function(redu) {
    title <- paste0(redu, "_vs_", grouping)
    p <- 
      monocle3::plot_cells(
        cds, 
        reduction_method = redu, 
        color_cells_by = grouping,
        show_trajectory_graph = F, 
        label_cell_groups = F) 
    if (!is.numeric(SummarizedExperiment::colData(cds)[, grouping])) {
      p <- p + ditto_colours
    }
    groupings_vs_reductions[[paste0(grouping, "_legend")]] <<-
      lemon::g_legend(p)
    groupings_vs_reductions[[title]] <<- 
      p + 
      ggplot2::theme(legend.position = "none")
  })
})
# create grob layout
p <-
  gridExtra::marrangeGrob(
    grobs = groupings_vs_reductions,
    nrow = length(avail_groupings),
    ncol = length(redus) + 1,
    layout_matrix = matrix(
      1:length(groupings_vs_reductions),
      length(avail_groupings),
      length(redus) + 1,
      TRUE
    ),
    top = NULL
  )
p
```

## Clustering

From [**Clarke et al., 2021**](https://www.nature.com/articles/s41596-021-00534-0)

>  ***Annotating cell states and gradients***
>
>  *When analyzing and characterizing novel cell types, it is important to determine whether they represent a stable cell type or contain multiple cell states. The definitions of cell type and state are not yet standardized, but a stable cell type may be expected to have homogeneous gene expression across a cluster and be compact in a 2D projection plot, whereas cell gradients appear as a spread-out string of cells and cell states (e.g., cell cycle state). Expression gradients indicate continuous differences that are present in the cell population, which could represent states like the cell cycle, immune activation63, spatial patterning64 or transient developmental stages. Care must be taken to distinguish biologically meaningful cell states and experimental batch effects, which can manifest in a similar way.*

We perform unsupervised transcriptome similarity-based clustering to identify groups of nuclei with relatively homogeneous transcript profiles (united by a shared celltype / state). Clustering is performed at different resolutions to explore how groupings change across the parameter space. The aim is to find a resolution at which the clusters appear to best reflect true biological subpopulations of the dataset. 

This process is based on the assumption that true discrete clusters of nuclei do exist within the dataset. The preceding nucleus filtering, feature selection, and dimensionality reduction steps were taken to distill only the highest quality features that reflect the underlying structure of the population and to remove noise that distracts from this structure.  

As with the final dimensionality parameter, there is no deterministic way to set the final clustering resolution and no definitive best value. Instead, one must qualitatively assess clustering outputs for the given dataset and pick a 'good' resolution. Typically, a 'good' final resolution is taken as one that groups nuclei into celltypes. However, looking at multiple clustering resolutions of the same dataset can provide multiple true and valuable conclusions about the dataset at multiple levels of functional subspecialisation of cells. 

Setting a **low clustering resolution** will produce fewer clusters, reflect the broadest subdivisions between groups of nuclei (e.g. different tissues / differentiation lineages), and is more robust to noise, but can miss important local structure within the dataset. 

**Higher clustering resolutions** will result in more, increasingly granular clusters and can reveal celltype-level groups. In can also reveal new / rare celltypes that are hidden at lower resolutions. Setting the clustering resolution too high, however, can produce meaningless groups and lead to overfitting. 

`monocle3` offers an unsupervised clustering function. This function also separates cells into superclusters, called partitions. These represent higher order communities within the dataset, allowing the different resolutions to coexist for downstream analysis.

From **`monocle3::cluster_cells` vignette**

> *Unsupervised clustering of cells is a common step in many single-cell expression workflows. In an experiment containing a mixture of cell types, each cluster might correspond to a different cell type. This function takes a cell_data_set as input, clusters the cells using Louvain/Leiden community detection, and returns a cell_data_set with internally stored cluster assignments. In addition to clusters this function calculates partitions, which represent superclusters of the Louvain/Leiden communities that are found using a kNN pruning method.*

From [**monocle3 documentation**](https://cole-trapnell-lab.github.io/monocle3/docs/trajectories/)

> *Although cells may continuously transition from one state to the next with no discrete boundary between them, Monocle does not assume that all cells in the dataset descend from a common transcriptional "ancestor". In many experiments, there might in fact be multiple distinct trajectories. For example, in a tissue responding to an infection, tissue resident immune cells and stromal cells will have very different initial transcriptomes, and will respond to infection quite differently, so they should be a part of the same trajectory.*
>
> *Monocle is able to learn when cells should be placed in the same trajectory as opposed to separate trajectories through its clustering procedure. Recall that we run cluster_cells(), each cell is assigned not only to a cluster but also to a partition. When you are learning trajectories, each partition will eventually become a separate trajectory. *

We can now cluster the cells. Monocle3 does not require a user-defined clustering resolution, as it can detect an optimum resolution automatically. Additionally, it groups the data into both high-resolution clusters and low-resolution superclusters (a.k.a. partitions) simultaneously.

```{r cluster_cells}
lvls <- c("cluster", "partition")
cds <- xfun::cache_rds({
  cds <- monocle3::cluster_cells(cds, random_seed = random_seed)
  # add it to the colData and reorder clusters and partitions numerically
  SummarizedExperiment::colData(cds)$cluster <- 
    factor(monocle3::clusters(cds), levels = sort(as.integer(levels(monocle3::clusters(cds)))))
  SummarizedExperiment::colData(cds)$partition <- 
    factor(monocle3::partitions(cds), levels = sort(as.integer(levels(monocle3::partitions(cds)))))
  cds
}, file = "cds_clustered.rds") 
```

```{r clustering_tree, class.source = 'fold-hide', fig.dim = c(10, 5), fig.cap = "Clustering tree from partitions (grouping 1) to clusters (grouping 2) of cells"}
library(clustree)
p <-
  data.frame("grouping1" = monocle3::partitions(cds),
             "grouping2" = monocle3::clusters(cds)) %>%
    clustree::clustree(prefix = "grouping")
p
```

```{r clusters_and_partitions, class.source = 'fold-hide', fig.dim = c(10, 5), fig.cap = "Plot of monocle3 UMAP, coloured by cluster and by partition"}
final_umap <- 
  lvls %>%
  purrr::map(function(lvl) {
    dittoSeq::dittoDimPlot(cds, lvl, size = 0.5, xlab = NULL, ylab = NULL) +
    ggplot2::geom_label(
      data = SummarizedExperiment::colData(cds) %>%
        dplyr::as_tibble() %>%
        dplyr::right_join(get_centroids(cds, "UMAP", lvl), by = lvl) %>%
        dplyr::distinct(get(lvl), x, y),
      ggplot2::aes(x, y, label = `get(lvl)`),
      size = 3, fill = "white", alpha = 0.7, label.size = NA) +
    umap_void_theme
  })  
names(final_umap) <- lvls
p <- 
  final_umap[["cluster"]] + final_umap[["partition"]]
p
```

## Differential expression analysis

Once cells have been clustered, we can isolate genes that are highly differentially expressed between the clusters. These cluster marker genes can indicate active programmes 

First, we identify the top genes that are most specifically expressed in each cluster and each partition. 

```{r cluster_markers, eval = T, results = F, warning = F, message = F}
# monocle3::top_markers has a bug, so this function doesn't work:
# marker_test_res <- monocle3::top_markers(cds, group_cells_by = "cluster", reference_cells = 1000)

# using a debugged version of the function ()
cds_markers <- xfun::cache_rds({
  cds_markers <- 
    lvls %>%
    purrr::map(function(lvl) {
      top_markers_develop(cds, group_cells_by = lvl, genes_to_test_per_group = 100)
    }) %>%
    setNames(lvls)
  cds_markers
}, file = "cds_markers.rds")

# get top 3 markers per cluster/partition
top_cds_markers <- 
  cds_markers %>%
  purrr::map(function(lvl) {
    lvl %>%
      dplyr::as_tibble() %>%
      dplyr::filter(fraction_expressing >= 0.10) %>%
      dplyr::group_by(cell_group) %>%
      dplyr::arrange(desc(as.numeric(cell_group))) %>%
      dplyr::top_n(3, pseudo_R2)
  })
readr::write_tsv(top_cds_markers %>% dplyr::bind_rows(.id = "cell_grouping"), 
                 "top_3_markers.tsv")
```

```{r top_markers, eval = T, class.source = 'fold-hide', fig.dim = c(10, dplyr::n_distinct(monocle3::clusters(cds))), fig.cap = "Plot of the mean expression across cells and percentage of cells in each group that express the top 3 marker genes of each group"}
# monocle3::plot_genes_by_group has a bug, so this function doesn't work:
# p <-
#   monocle3::plot_genes_by_group(
#     cds,
#     markers = top_specific_marker_ids,
#     ordering_type = "maximal_on_diag",
#     max.size=3) +
#     ggplot2::theme(text = ggplot2::element_text(size = 8))

# using dittoSeq::dittoDotPlot instead
p <- 
  lvls %>%
  purrr::map(function(lvl) {
    dittoSeq::dittoDotPlot(
      cds, 
      vars = top_cds_markers[[lvl]] %>% dplyr::pull(gene_id) %>% unique(),
      group.by = lvl,
      x.labels.rotate = F) +
    ggplot2::coord_flip() +
    ggplot2::theme(text = ggplot2::element_text(size = 8))
  }) %>%
  gridExtra::marrangeGrob(nrow = 1, ncol = 2, top = NULL)
p
```

## Celltype annotation

The conventional approach to celltype annotation for sc/snRNAseq datasets is to inspect known marker genes from each of the celltypes that are expected and likely to be found at the sample site. Known relationships between marker genes and celltypes of interest can be obtained from databases or manually curated from literature. Then, the expression profiles of all markers from all celltypes across all clusters is programmatically or visually inspected and clusters are labelled according to their marker profile. 

### Automated celltype annotation using database markers

The package `SingleR` returns "the best annotation for each cell in a test dataset, given a labelled reference dataset in the same feature space". Using the human primary cell atlas from `celldex`, we can find the nearest reference celltype to each nucleus in the dataset. This approach is good because it is unbiased and does not require manual assignment, but it is unclear how well snRNAseq output will map onto a scRNAseq atlas. Additionally, the reference dataset from `celldex` is more rigid, cannot be customised to the use case, and does not contain transformed celltypes.

Some annotations will contain very few nuclei. These are usually not of interest and clutter up plotting, so we remove annotations assigned to < 10 nuclei.

```{r singler_annots, message = F, warning = F}
cds_singler <- xfun::cache_rds({
  singler_refs <- celldex::BlueprintEncodeData()
  cds_singler <- SingleR::SingleR(
    test = cds@assays@data$counts,
    ref = singler_refs,
    labels = colnames(singler_refs))
  cds_singler
}, file = "cds_singler.rds")

cds <- xfun::cache_rds({
  # add singler annotation
  SummarizedExperiment::colData(cds) <- 
    SummarizedExperiment::colData(cds) %>%
    cbind(
      cds_singler %>%
    dplyr::as_tibble() %>%
    dplyr::select(singler_annot_fine = labels) %>%
    dplyr::group_by(singler_annot_fine) %>%
    dplyr::left_join(
      SummarizedExperiment::colData(singler_refs) %>%
        tibble::as_tibble(rownames = "singler_annot_fine") %>%
        dplyr::select(singler_annot_fine, singler_annot_main = label.main))
    )
  # add main/fine labels for partitions/clusters
  for (lvl in c("cluster", "partition")) {
    for (annot in c("singler_annot_fine", "singler_annot_main")) {
      SummarizedExperiment::colData(cds)[, paste(lvl, annot, sep = "_")] <-
        get_singler_annot_label(cds, annot, lvl)
    }
  }
  # return
  cds
}, file = "cds_singler_annotated.rds")
```

```{r singler_annot_fine_heatmap, class.source = 'fold-hide', fig.width = 10, fig.cap = "Heatmap of fine celltype annotation scores from SingleR"}
annotation_col <- 
  SummarizedExperiment::colData(cds)[, c("cluster", "partition", "singler_annot_main")] %>% 
  as.data.frame()
if (dplyr::n_distinct(SummarizedExperiment::colData(cds)$nih_pid, na.rm = T) > 1) {
  annotation_col$nih_pid <- SummarizedExperiment::colData(cds)$nih_pid
}
if (dplyr::n_distinct(SummarizedExperiment::colData(cds)$lesion_type, na.rm = T) > 1) {
  annotation_col$lesion_type <- SummarizedExperiment::colData(cds)$lesion_type
}
cds_singler$main_labels <- SummarizedExperiment::colData(cds)$singler_annot_main
p <-
  SingleR::plotScoreHeatmap(
    cds_singler, 
    show.labels = F,
    annotation_col = annotation_col,
    fontsize_row = 7)
p
```

```{r umap_vs_singler_annots, class.source = 'fold-hide', results = F, fig.width = 10, fig.height = 10, fig.cap = "Plot of UMAP clusters, coloured by SingleR annotation and labelled by the major SingleR annotation(s) per cluster"}
p <- list()
purrr::walk(c("cluster", "partition"), function(lvl) {
  purrr::walk(c("singler_annot_fine", "singler_annot_main"), function(annot) {
    label_col <- paste(lvl, annot, sep = "_")
    p_i <-
      dittoSeq::dittoDimPlot(cds, annot, main = gsub("_", " ", label_col), size = 0.7) +
      ggrepel::geom_label_repel(
        data = 
          dplyr::as_tibble(SummarizedExperiment::colData(cds)) %>%
          dplyr::select(dplyr::all_of(c(lvl, label_col))) %>%
          dplyr::right_join(get_centroids(cds, "UMAP", lvl), by = lvl) %>%
          dplyr::distinct(),
        ggplot2::aes(x, y, label = get(label_col)),
        min.segment.length = 0,
        size = 1.5, fill = 'white', alpha = 0.8, seed = 1234, label.size = NA
      ) 
    p[[label_col]] <<- p_i + umap_void_theme 
  })
}) 
patchwork::wrap_plots(p)
```

```{r singler_annots_bar, class.source = 'fold-hide', fig.cap = "Plot of SingleR annotation composition of samples"}
p <-
  dittoSeq::dittoBarPlot(
  cds,
  var = "singler_annot_main",
  group.by = "sample",
  scale = "count",
  main = "",
  legend.show = F
  ) +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, size = 5),
                 legend.text = ggplot2::element_text(size = 5)) +
  SummarizedExperiment::colData(cds) %>%
  tibble::as_tibble() %>%
  ggplot2::ggplot(ggplot2::aes(x = cluster, fill = singler_annot_main)) +
  ggplot2::geom_bar() + 
  ditto_colours +
  ggplot2::theme_classic() +
  ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 90, size = 5),
                 legend.text = ggplot2::element_text(size = 5),
                 legend.title = ggplot2::element_blank()) +
  patchwork::plot_layout(
    ncol = 2,
    widths = c(
      SummarizedExperiment::colData(cds)$sample %>% dplyr::n_distinct(),
      SummarizedExperiment::colData(cds)$cluster %>% dplyr::n_distinct()
    )
  )
p
```

```{r singler_annots_alluvial, class.source = 'fold-hide', fig.width = 10, fig.cap = "Alluvial plot of assignment paths of nuclei from samples to clusters to partitions to SingleR annotations, coloured by cluster"}
# TODO: # fix this
p <-
  dplyr::as_tibble(SummarizedExperiment::colData(cds)) %>%
  dplyr::group_by(sample, partition, cluster, singler_annot_main) %>%
  dplyr::count() %>%
  dplyr::ungroup() %>%
  ggforce::gather_set_data(1:4) %>%
  dplyr::mutate(x = dplyr::case_when(x == 1 ~ "sample",
                                     x == 2 ~ "partition",
                                     x == 3 ~ "cluster",
                                     x == 4 ~ "singler_annot_main") %>%
                  factor(levels = c("sample", "partition", "cluster", 
                                    "singler_annot_main"))) %>%
  ggplot2::ggplot(ggplot2::aes(
                    x = x,
                    id = id,
                    split = y,
                    value = n
                  )) +
  ggforce::geom_parallel_sets(ggplot2::aes(fill = cluster),
                              axis.width = 0.15,
                              alpha = 0.75) +
  ggforce::geom_parallel_sets_axes(fill = "lightgrey",
                                   axis.width = 0.15) +
  ggforce::geom_parallel_sets_labels(angle = 0, 
                                     nudge_x = c(
                                       rep(-0.3, dplyr::n_distinct(SummarizedExperiment::colData(cds)$sample)),
                                       rep(0, dplyr::n_distinct(SummarizedExperiment::colData(cds)$partition)),
                                       rep(0, dplyr::n_distinct(SummarizedExperiment::colData(cds)$cluster)),
                                       rep(0.4, dplyr::n_distinct(SummarizedExperiment::colData(cds)$singler_annot_main))
                                     )) +
  ggplot2::theme_void() +
  ggplot2::theme(legend.position = "none",
                 axis.text.x = ggplot2::element_text(face = "bold", size = 15)) +
  ditto_colours
p
```

### Manual celltype annotation using curated literature markers

`r if (curr_params$input$markers_file == F) {"No markers_file given." }`

```{r exit_if_no_markers, eval = curr_params$input$markers_file == F}
knitr::knit_exit()
```

```{r cds_avail_markers, echo = F}
markers <-
  curr_params$input$markers_file %>%
  readr::read_tsv()
markers <- 
  split(markers$gene, markers$population)
avail_markers <- markers %>% purrr::map(intersect, rownames(cds))
```

Literature markers for celltypes of interest in VHL ccRCC tumours were gathered from previous sc/snRNAseq studies of sporadic ccRCC. 

```{r all_modules_heatmap, echo = F, warning = F, message = F, fig.dim = c(10, 20), fig.cap = "Heatmap of all module scores for all monocle3 clusters"}
df <- 
  avail_markers %>%
  purrr::map(dplyr::as_tibble) %>%
  dplyr::bind_rows(.id = "module") %>%
  dplyr::distinct() %>%
  dplyr::mutate(unique_module = module) %>%
  dplyr::group_by(unique_module) %>%
  dplyr::mutate(module_size = dplyr::n_distinct(value),
                module = paste0("(", module_size, ") ", module)) %>%
  dplyr::ungroup()

# genes, grouped into modules
gene_group_df <- 
  df %>% 
  dplyr::select(value, unique_module) 

# cells, grouped into clusters and partitions
cell_group_df <- 
  tibble::tibble(
    cell = row.names(SummarizedExperiment::colData(cds)), 
    cluster = monocle3::clusters(cds),
    partition = monocle3::partitions(cds))

# col annotations (cluster, partition)
col_annotations <-
  cell_group_df %>%
  dplyr::distinct(cluster, partition) %>%
  tibble::column_to_rownames("cluster")

# get row and col annotation colours
colours <- 
  list(partition = col_annotations$partition) %>%
  purrr::map(function(i) {
    dittoSeq::dittoColors()[1:length(unique(i))] %>%
      setNames(unique(i))
    }
  )

# check if >1 partition
multiple_partitions <- dplyr::n_distinct(cell_group_df$partition) > 1
multiple_clusters <- dplyr::n_distinct(cell_group_df$cluster) > 1
```

#### Consensus modules vs clusters

```{r modules_vs_clusters_heatmap, eval = multiple_clusters == T, echo = F, warning = F, message = F, fig.dim = c(10, 20), fig.cap = "Heatmap of all module scores for all monocle3 clusters"}
agg_mat <- 
  monocle3::aggregate_gene_expression(
    cds, 
    gene_group_df, 
    cell_group_df[, c("cell", "cluster")])
p <-
  pheatmap::pheatmap(agg_mat, scale = "row", clustering_method = "ward.D2",
                     annotation_col = col_annotations,
                     annotation_colors = colours, border_color = NA)
p
```

#### Consensus modules vs partitions

```{r modules_vs_partitions_heatmap, eval = multiple_partitions == T, echo = F, warning = F, message = F, fig.dim = c(10, 15), fig.cap = "Heatmap of all module scores for all monocle3 partitions"}
agg_mat <- 
  monocle3::aggregate_gene_expression(
    cds, 
    gene_group_df, 
    cell_group_df[, c("cell", "partition")])
p <-
  pheatmap::pheatmap(agg_mat, scale = "row", clustering_method = "ward.D2",
                     annotation_colors = colours, border_color = NA)
p
```

#### HIF target gene expression

```{r hif_genes_vs_cells, eval = multiple_partitions == T, echo = F, warning = F, message = F, fig.dim = c(10, 10), fig.cap = "Plot of HIF module score for all cells"}
p <- monocle3::plot_cells(cds, genes = avail_markers$HIF, label_groups_by_cluster = F)
p
```

```{r knit_module_plots, echo = FALSE, warning = F, message = F, results = 'asis'}
# get available markers

to_knit <- c()
agg_mat <- list()
genes <- list()
cell_group_df <- list()


for(ct in names(avail_markers)) {
  
  to_knit <- c(to_knit, paste0("\n##### Population: **", ct, "**\n"))
  chunk_name_ct <- ct %>% janitor::make_clean_names()
  ct_markers <- avail_markers[[ct]]
  n_ct_markers <- length(ct_markers)
  
  for(lvl in lvls) {

    chunk_name_lvl <- paste(chunk_name_ct, lvl) %>% janitor::make_clean_names()

    if (dplyr::n_distinct(SummarizedExperiment::colData(cds)[, lvl]) > 1 & 
        n_ct_markers > 1) {
      
      genes[[chunk_name_lvl]] <- avail_markers[[ct]]
      cell_group_df[[chunk_name_lvl]] <-
        tibble::tibble(cell = row.names(SummarizedExperiment::colData(cds)),
                       cell_group = SummarizedExperiment::colData(cds)[, lvl],
                       partition = monocle3::partitions(cds))
      agg_mat[[chunk_name_lvl]] <-
        monocle3::aggregate_gene_expression(
          cds[genes[[chunk_name_lvl]], ],
          cell_group_df = cell_group_df[[chunk_name_lvl]][, c("cell", "cell_group")])
      knit_lvl_module_heatmap <- c(
        '',
        paste0('```{r ', chunk_name_lvl, '_gene_heatmap, class.source = "fold-hide",  warning = F, message = F, fig.dim = c(10, max(length(genes[[chunk_name_lvl]]) / 6, 2))}'),
        paste0('p <- pheatmap::pheatmap(agg_mat[["', chunk_name_lvl, '"]], scale = "row", clustering_method = "ward.D2", fontsize = 6', ifelse(n_ct_markers == 1, ", cluster_rows = F", ""), ')'),
        'p',
        '```',
        ''
        )
      } else {
        knit_lvl_module_heatmap <- paste0(
          "Only one ", ifelse(n_ct_markers == 1, "gene", lvl), " so cannot produce a heatmap!\n")
      }
    
    dev_off_if()
    knitted_lvl_module_heatmap <- knitr::knit_child(text = knit_lvl_module_heatmap, envir = environment(), quiet = TRUE)
    dev_off_if()
    to_knit <- c(to_knit, knitted_lvl_module_heatmap)
  
  }

  if(n_ct_markers < 50) {
    knit_module_dimplots <- c(
      '',
      paste0('```{r ', chunk_name_ct, ', echo = F, warning = F, message = F, fig.dim = get_fig_dims((n_ct_markers + 2), n_cols = 4, height_to_width_ratio = 1.2)}'),
      'p <- plot_markers_on_umap(cds, ml = ct_markers, final_umap)',
      'p',
      '```',
      ''
      )
  } else {
    knit_module_dimplots <- paste(n_ct_markers, "markers in this set - too many to print!\n")
  }

  dev_off_if()
  knitted_module_dimplots <- knitr::knit_child(text = knit_module_dimplots, envir = environment(), quiet = TRUE)
  dev_off_if()
  to_knit <- c(to_knit, knitted_module_dimplots)

}
dev_off_if()
cat(unlist(to_knit), sep = '\n')
```

## Final cluster annotations

```{r summarise_annots, echo = F}
cell_groupings_summary <- xfun::cache_rds({
  
  cell_groupings <-
    dplyr::as_tibble(SummarizedExperiment::colData(cds)) %>%
    tidyr::pivot_longer(c("cluster", "partition"), names_to = "cell_grouping", values_to = "cell_group")
  
  # counts
  cell_groupings_summary <-
    cell_groupings %>% 
    dplyr::group_by(cell_grouping, cell_group) %>%
    dplyr::count() 
  
  # add percent by lesion type
  if ("lesion_type" %in% names(cell_groupings)) {
    cell_groupings_summary <-
      cell_groupings_summary %>%
      dplyr::left_join(
      cell_groupings %>%
        dplyr::group_by(cell_grouping, cell_group, lesion_type) %>%
        dplyr::count() %>% 
        dplyr::group_by(cell_grouping, cell_group) %>%
        dplyr::mutate(prop = n / sum(n)) %>%
        tidyr::pivot_wider(names_from = lesion_type, values_from = prop, id_cols = dplyr::starts_with("cell_group")),
      by = c("cell_grouping", "cell_group")
    ) 
  }
  
  # add percent by singler annot
  cell_groupings_summary <-
    cell_groupings_summary %>%
    dplyr::left_join(
      # singler_annots
      cell_groupings %>%
        tidyr::pivot_longer(dplyr::starts_with("singler_annot_")) %>%
        dplyr::group_by(cell_grouping, cell_group, name, value) %>%
        dplyr::count() %>% 
        dplyr::group_by(cell_grouping, cell_group, name) %>%
        dplyr::mutate(prop = n / sum(n) * 100,
                      label = paste0(value, " (", round(prop, ifelse(prop < 1, 1, 0)), "%)")) %>% 
        dplyr::filter(prop > 10) %>%
        dplyr::arrange(-prop) %>%
        dplyr::summarise(value = paste(label, collapse = ", ")) %>%
        tidyr::pivot_wider(),
      by = c("cell_grouping", "cell_group")
    ) %>%
    dplyr::left_join(
      # top_cds_markers
      top_cds_markers %>%
        dplyr::bind_rows(.id = "cell_grouping") %>%
        dplyr::group_by(cell_grouping, cell_group) %>% 
        dplyr::summarise(top_markers = paste(gene_id, collapse = ", ")),
      by = c("cell_grouping", "cell_group")
    )
  
  cell_groupings_summary
  
}, file = "cell_groupings_summary.rds")

knitr::kable(cell_groupings_summary, digits = 3)
```